internal/slicer/slicer.go

```diff
@@ -397,3 +391,63 @@ func createFile(targetPath string, pathInfo setup.PathInfo) (*fsutil.Entry, erro
 		MakeParents: true,
 	})
 }
+
+// selectPkgArchives selects the appropriate archive for each selected slice
+// package. It returns a map of archives indexed by package names.
+func selectPkgArchives(archives map[string]archive.Archive, selection *setup.Selection) (map[string]archive.Archive, error) {
+	// Sort the archives by their priority value.
+	releaseArchives := make([]*setup.Archive, 0, len(selection.Release.Archives))
+	for _, archive := range selection.Release.Archives {
+		if archive.Priority < 0 {
+			// Ignore negative priority archives unless a package specifically
+			// asks for it with the "archive" field.
+			continue
+		}
+		releaseArchives = append(releaseArchives, archive)
+	}
+	slices.SortFunc(releaseArchives, func(a, b *setup.Archive) int {
+		if a.Priority > b.Priority {
+			return -1
+		}
+		if a.Priority < b.Priority {
+			return 1
+		}
+		return 0
+	})
+
+	// Select the appropriate archive for each slice package.
+	pkgArchives := make(map[string]archive.Archive)
+	for _, s := range selection.Slices {
+		pkg := selection.Release.Packages[s.Package]
+		if _, ok := pkgArchives[pkg.Name]; ok {
+			continue
+		}
+		if pkg.Archive == "" {
+			// If the package has not pinned any archive, choose the highest
+			// priority archive in which the package exists.
+			var chosen archive.Archive
+			for _, releaseArchive := range releaseArchives {
+				archive := archives[releaseArchive.Name]
+				if archive != nil && archive.Exists(pkg.Name) {
+					chosen = archive
+					break
+				}
+			}
+			if chosen == nil {
+				return nil, fmt.Errorf("slice package %q missing from archive(s)", pkg.Name)
+			}
+			pkgArchives[pkg.Name] = chosen
+		} else {
+			// If the package has pinned a valid archive, select that one.
+			archive := archives[pkg.Archive]
```

Comment:
There should be a way of unifying the logic here. This is just an idea I tested locally to see if it made sense, tell me what you think:
```go
	// Sort the archives by their priority value.
	sortedArchives := make([]*setup.Archive, 0, len(selection.Release.Archives))
	for _, archive := range selection.Release.Archives {
		if archive.Priority < 0 {
			// Ignore negative priority archives unless a package specifically
			// asks for it with the "archive" field.
			continue
		}
		sortedArchives = append(sortedArchives, archive)
	}
	slices.SortFunc(sortedArchives, func(a, b *setup.Archive) int {
		return b.Priority - a.Priority
	})

	// Select the appropriate archive for each slice package.
	pkgArchives := make(map[string]archive.Archive)
	for _, s := range selection.Slices {
		pkg := selection.Release.Packages[s.Package]
		if _, ok := pkgArchives[pkg.Name]; ok {
			continue
		}

		// If the package has not pinned any archive, choose the highest
		// priority archive in which the package exists.
		var sortedCandidates []*setup.Archive
		if pkg.Archive != "" {
			sortedCandidates = []*setup.Archive{selection.Release.Archives[pkg.Archive]}
		} else {
			sortedCandidates = sortedArchives
		}

		var chosen archive.Archive
		for _, archiveInfo := range sortedCandidates {
			archive := archives[archiveInfo.Name]
			if archive != nil && archive.Exists(pkg.Name) {
				chosen = archive
				break
			}
		}
		if chosen == nil {
			return nil, fmt.Errorf("cannot find package %q in archive", pkg)
		}
		pkgArchives[pkg.Name] = chosen
	}
	return pkgArchives, nil
```

You can do it in any other way, the idea was to have a list of archives and then see if the package is there. When there is archive pinning we use the same list but there is only one element. This will also come in handy if/when we allow package.Archive to be a list in the future.

Reply:
To be honest, I would much rather keep it separate. Mostly because of this one below:

```go
			archive := archives[pkg.Archive]
			if archive == nil {
				return nil, fmt.Errorf("archive %q not defined", pkg.Archive)
			}
```

I would like the error to be reported if a pinned archive actually does not exist and produce a different error msg. But if you do feel strongly about unifying, we can go that way. I don't have a very strong preference towards keeping the logic separate, only slightly.

Reply:
We will unify it and move the check I mentioned above inside `release.validate()` in `setup.go`.

Reply:
This is done in https://github.com/letFunny/chisel/pull/12/commits/1f4a4545e7dfbde0d5309b2a3ed4474a28afd460.